### ISSUE TITLE
Usage of gstatic.com and CDN violates German law

### ISSUE THREAD

**Issue by jans23** [odoo/odoo#83638]
Impacted versions: 12.0-15.0 (at least)

Steps to reproduce: Open the network monitor in your browser. Then open any version of Odoo at runbot.odoo.com and see connections to Google and Cloudflare.

Current behavior: 

Content is downloaded from:
fonts.gstatic.com
fonts.googleapis.com
www.google.com
cdnjs.cloudflare.com

At least [according to German law based on GDPR](https://rewis.io/urteile/urteil/lhm-20-01-2022-3-o-1749320/) using third party websites such as Google and Cloudflare is problematic because visitors' IP address is a personal information

Expected behavior:

It should be possible to configure Odoo in a way that content is not downloaded from third party but hosted locally.

**Comment by odony** [comment]
Hello,

To clarify, this is not a German law, it is a single court ruling, saying in a nutshell that in this case, the court considered that using the "Legitimate Interest" basis (Art. 6.1 (f) GDPR) for the processing of visitor IPs (and the transfer of this info to Google Fonts) was not valid. The fine that imposed is rather symbolic at this point.

This means that, for example, a website owner could still use Google Fonts, but they would have to rely on a different legal basis, such as Consent (Art. 6.1 (a)). As it's commonly done for cookies, this would require asking permission from the user before processing, e.g. with a consent pop-up. That's an important distinction to make.

As for customizing fonts, Odoo includes a powerful editor and theming system that allows overriding styles easily, including fonts. If some users are concerned about this, I'm sure @qsm-odoo or the @odoo/rd-website team can show an elegant customisation to accomplish it.

BTW I don't believe the backend uses Google Fonts at all because it's not supposed to be customized a lot on that front, this is mainly a feature for websites.

**Comment by qsm-odoo** [comment]
@odony We are already aware of the issue via https://www.odoo.com/web#id=2756486&cids=1&model=project.task&view_type=form we still have to discuss more about it :+1: 

cc @rdeodoo

**Comment by jans23** [comment]
@qsm-odoo Any news on this?

**Comment by rdeodoo** [comment]
There is a WIP being developed internally to fetch/download google fonts server side and store it in an attachment that will be served instead of the google font.
Note that this is not a top priority right now.

**Comment by jaykay** [comment]
Hi, we got an Odoo Enterprise customer who got a cease and desist letter because of that issue. Would be great if the solution could be expedited.

**Comment by odony** [comment]
> we got an Odoo Enterprise customer who got a cease and desist letter because of that issue. 

Hello, if a company actually got a C&D letter:
- First, they should get in touch with trustworthy legal advisors, to verify the legal grounds for whatever claims are in there. The majority of C&D letters are just attempts at bullying people, without any legal merit.
- Then, if a verifiable reason to comply is found by legal advisors, get in touch with the Odoo Helpdesk to ask for one of the possible workarounds.

Website fonts are one of the numerous things that are meant to be customizable in Odoo, and there is no need to wait for a future update, if you have an immediate solution.

PS: there's also no guarantee that we will make any change for this other than making sure users can easily customize fonts, and switch to local ones in *some way*. Which should already be possible - but maybe not easily enough.

**Comment by jaykay** [comment]
Hi @odony now we have at least 3 Enterprise customers who asked us to make sure to avoid that issue. To be exactly clear: Using Odoo as a website builder in Germany is illegal since January 2022 due to the way Odoo uses Google Fonts. As this is not configurable for the customers, they are not able to fulfill the law. There's a bunch of lawyers right now underway to send out c&d letters to anyone using Google Fonts from gstatic.
In Germany it is everywhere in the media. Now the question is: Will Odoo comply with German law? Or should we tell the customers to avoid using Odoo for websites which would result in some of Odoos major benefits and USPs to be useless. (At least in Germany)

**Comment by qsm-odoo** [comment]
@jaykay We are not forgetting this. We'll definitely provide a solution (at least in the form of an alternative for impacted users), that is still being debated. Currently the website builder team has its hands full with the upcoming release of version 16.0 but we will definitely tackle this problem in the upcoming weeks, I'd say by the end of the month.

**Comment by EwetoyeIbrahim** [comment]
> Hello,
> 
> To clarify, this is not a German law, it is a single court ruling, saying in a nutshell that in this case, the court considered that using the "Legitimate Interest" basis (Art. 6.1 (f) GDPR) for the processing of visitor IPs (and the transfer of this info to Google Fonts) was not valid. The fine that imposed is rather symbolic at this point.
> 
> This means that, for example, a website owner could still use Google Fonts, but they would have to rely on a different legal basis, such as Consent (Art. 6.1 (a)). As it's commonly done for cookies, this would require asking permission from the user before processing, e.g. with a consent pop-up. That's an important distinction to make.
> 
> As for customizing fonts, Odoo includes a powerful editor and theming system that allows overriding styles easily, including fonts. If some users are concerned about this, I'm sure @qsm-odoo or the @odoo/rd-website team can show an elegant customisation to accomplish it.
> 
> BTW I don't believe the backend uses Google Fonts at all because it's not supposed to be customized a lot on that front, this is mainly a feature for websites.

Off my head, simply raising a notification on website visit just as commonly used for cookies should be sufficient.

**Comment by rdeodoo** [comment]
The thing is, the visitor would then have already pinged the google servers, which seems to be the problematic part here: google knows who you are and that you visited that website.

And if we were to only load the actual google font once the user agrees the cookie policy, it means that every access to the website would be with a default font (while waiting for the consent). Then using a fond when building a website would not have a lot of sense.

We still didn't forget about this one.
We tried a first quick and naive approach internally a few weeks ago which didn't go further.

We plan to ship another opt-in solution in stable before the release of the v16 (so it's part of v16): we will add an unchecked checkbox when uploading a google font in the right panel of the website builder, which will be something like "Download and serve locally instead of from the Google server".
That way, people who really want a workaround will be able to.

**Comment by rdeodoo** [comment]
Took a bit longer than expected but a solution is now available in 16.0 with https://github.com/odoo/odoo/commit/b06ce21eba6388ce34bbffffadcb489f0e8557dd

We will backport it in 14.0 in a few weeks, once it lived a bit in 16.0, better safe than sorry in case we missed something.

I will close this issue once it has been backported.

**Comment by jans23** [comment]
That is great. Thank you so much. Do you plan to backport it to 15.0 too?

**Comment by rdeodoo** [comment]
Yes of course, `We will backport it in 14.0` means 14.0 and later versions :)

**Comment by ap-wtioit** [comment]
The odoo online instances set up after the launch of 16.0 (on thursday of the odoo experience) still include google fonts directly loaded from google and thus sharing the visitors data with google without consent.

**Comment by rdeodoo** [comment]
Not by default.
The default font is now the system one, letting your system decide which **local** font to use.
From there, if you want to use another google font, you can either:
- Select one of the (unselected) suggested / listed fonts in the dropdown, that will use google server to serve
- Add a new one yourself from google fonts and let the checkbox/toggle set to "Let google serve it", that will use google server to serve
- Add a new one yourself from google fonts and uncheck the  "Let google serve it" checkbox/toggle, that won't use google server but let your DB serve it as a locally downloaded attachment.

Note that most people still want to serve the fonts through google, it has multiple advantages:
- Up to date font (receive fix and imp, shouldn't be frequent though)
- More accurate font choice depending of the system: woff, woff2, eot, ttf, svg etc
- Probably faster, google is of course reliable and can handle huge workload, their infrastructure will always be better than yours
- Reduce the charge of your server (or our server if saas). Imagine millions of font call, it would be better letting those handled by google
- ..

That's why the "Let google serve it" toggle is ON by default (you have to uncheck it) and also why the provided few fonts options in the select are served through google.

**Comment by ap-wtioit** [comment]
@rdeodoo tecnicaly you might be right, when installing website the font is set to: `'font': 'SYSTEM_FONTS'` but that does not include `navbar-font` nor `headings-font`.

however as soon as the user finishes the website configurator and choosing a "beautiful" theme all bets are off and google fonts are chosen as default (with urls specified which makes them loading the data directly from google).

when searching through the themes i could not find a single one that uses SYSTEM_FONTS or local google font loading as default.

in theme_odoo_experts you cannot even set navbar-font in the theme editor (only font and headings-font) so it is still loading "Work Sans" from google after the user has done everthing he can (assuming he is not aware of how to change the font's in the database directly) to avoid that.

**Comment by rdeodoo** [comment]
> 'font': 'SYSTEM_FONTS' but that does not include navbar-font nor headings-font.

They now fallback to the `font` value and don't need to be explicitly provided, if I am not missing anything.

>  choosing a "beautiful" theme all bets are off and google fonts are chosen as default

About the themes, you are right, but we don't want to change the way our themes works:
- We obviously don't want to use the system font, that would kill the theme purpose
- We don't want to serve google font locally by default, for the reason I listed above

We provided those 2 solutions as a workaround / to help, really:
- Capability to locally host the fonts
- System font by default if no theme

If someone is concerned about the google font server and is using a theme, he should simply upload the font used by the theme himself and unselect the "Use Google Server" option. It will stop using the one of google server but use the local attachment one.
(Note that there will be a visual glitch in the right panel font selection dropdown if a same font is both used as local and remote, but we have a fix for that in 14.0, as it was also an issue there is you upload 2 times the same font)
